### PR TITLE
Bring StorageObjectInUseProtection feature to GA

### PR DESCRIPTION
**Pull request by pospispa**
StorageObjectInUseProtection is Beta in K8s 1.10.

It will be brought to GA in K8s 1.11 in PR https://github.com/kubernetes/kubernetes/pull/62870 that is not merged yet.

### CODE REVIEW

File: docs/tasks/administer-cluster/pvc-protection.md [hunk 1]
```diff
@@ -1,144 +0,0 @@
----
-reviewers:
-- msau42
-- jsafrane
-title: Persistent Volume Claim Protection
```
Comment (tengqm) [inline docs/tasks/administer-cluster/pvc-protection.md]: Should we just revise this task doc instead of removing it?

**Comment by pospispa** [comment]
/hold cancel
as the source code PR https://github.com/kubernetes/kubernetes/pull/62870 got merged.

File: docs/tasks/administer-cluster/pvc-protection.md (hunk 1 above)
Comment (pospispa) [reply docs/tasks/administer-cluster/pvc-protection.md]: The PR https://github.com/kubernetes/website/pull/7620 renamed the `docs/tasks/administer-cluster/pvc-protection.md` file to `docs/tasks/administer-cluster/storage-object-in-use-protection.md` and forgot to remove the `docs/tasks/administer-cluster/pvc-protection.md` file that is not used anywhere in the docs. So removal of the `docs/tasks/administer-cluster/pvc-protection.md` file is a fix of the PR https://github.com/kubernetes/website/pull/7620
IMHO, I should have created a separate PR for the fix (`docs/tasks/administer-cluster/pvc-protection.md` file removal) and not mix it with this PR.

Comment (tengqm) [reply docs/tasks/administer-cluster/pvc-protection.md]: Okay. Thanks for the clarification.

File: docs/tasks/administer-cluster/storage-object-in-use-protection.md [hunk 2]
```diff
@@ -6,7 +6,6 @@ title: Storage Object in Use Protection
 ---
 
 {% capture overview %}
-{% assign for_k8s_version="v1.10" %}{% include feature-state-beta.md %}
 
 Persistent volume claims (PVCs) that are in active use by a pod and persistent volumes (PVs) that are bound to PVCs can be protected from pre-mature removal.
```
Comment (Bradamant3) [inline docs/tasks/administer-cluster/storage-object-in-use-protection.md]: As long as we're making making changes to the file anyway:

s/pre-mature/premature

File: docs/admin/admission-controllers.md [hunk 3]
```diff
@@ -568,8 +568,7 @@ This admission controller will deny any pod that attempts to set certain escalat
 This admission controller implements automation for [serviceAccounts](/docs/user-guide/service-accounts).
 We strongly recommend using this admission controller if you intend to make use of Kubernetes `ServiceAccount` objects.
 
-### Storage Object in Use Protection (beta)
-{% assign for_k8s_version="v1.10" %}{% include feature-state-beta.md %}
+### Storage Object in Use Protection
```
Comment (Bradamant3) [inline docs/admin/admission-controllers.md]: Do we want to mention that this is a feature formerly known as PVC Protection? Or that it's called PVC Protection in earlier versions of K8s?

Comment (pospispa) [reply docs/admin/admission-controllers.md]: Firstly, IMHO, users of K8s 1.11 documentation are not interested in K8s history, I think they are only interested in K8s 1.11.
I believe that users interested in history, i.e. differences between particular K8s versions will read release notes of these K8s versions.
That's why I think we don't need to mention that `Storage Object in Use` feature was in K8s 1.9 implemented only partially and was called `PVC Protection`.

---

Lastly, @Bradamant3 your question:
> Same perhaps for the related tasks topic?

Makes me think of the following: `Storage Object In Use` feature currently protects PVCs and PVs. It may protect other Storage Objects in future. Is this clear from the documentation?

File: docs/tasks/administer-cluster/storage-object-in-use-protection.md (hunk 2 above)
Comment (pospispa) [reply docs/tasks/administer-cluster/storage-object-in-use-protection.md]: @Bradamant3 I added feature state code. Is it what you were looking for?

I intentionally skipped K8s 1.9, because `Storage Object In Use` feature didn't exist in K8s 1.9. The `Storage Object In Use` feature went directly to beta.

**Comment by Bradamant3** [comment]
@pospispa sorry for the delay on this one. I'm leaving it open so as not to confuse things (we've closed a bunch of PRs in similar state). 

We've migrated the site's framework to Hugo. Can you please change this PR to move files to their post-migration location (/content/en/docs/...) and reference this PR in the new PR's description? (You'll need to rebase the PR too for the new location to make sense, but you know that ...) Thank you! Any questions, please ping me.

Then I can approve and get it merged.

**Comment by mistyhacks** [comment]
The underlying structure of the docs repository has changed due to the recent
Hugo migration. Specifically, all the files previously under `docs/` are now in
`content/en/docs/`. I have rebased your PR without changing any of your work, so
that your changes are modifying the affected files in the new location, so this
PR is now reviewable again.

**If you were using Github's web UI rather than a local Git client, you can stop
reading and continue working in this PR.**

If you use the Git command-line client, you will need to make sure your local
branch is up to date with your fork. **If you have local commits that have not
been pushed to your fork, it may be complicated to get them back.** Do `git log`
and note their commit hashes for later.

Assuming your your fork's remote is called `origin`, hard-reset your local
branch to be the same as your fork's branch:

```bash
git fetch origin; git reset --hard origin/Bring-StorageObjectInUseProtection-feature-to-GA
```

**If you didn't have local commits, you're now ready to keep working on this PR
locally. You can stop reading now.**

If you did have local commits, cherry-pick them from oldest to newest. **For
each commit that needs to be cherry-picked:**

```bash
git cherry-pick <HASH>
```

You will need to resolve conflicts in these cherry-picked commits, which
involves moving your modified files into the correct new location. All the
content in `docs/` moved to `content/en/docs/`, so you can move your local files
accordingly:

- Do `git add` for each of the moved files
- Do `git mergetool` and accept the
deleted rather than modified file
- Continue the cherry-pick by doing `git cherry-pick --continue`


When you've cherry-picked all local commits back onto the local branch, push the
branch (with new commits) to your fork. You should not need to force the push.

File: content/en/docs/admin/admission-controllers.md [hunk 4]
```diff
@@ -21,7 +22,7 @@ is authenticated and authorized.  The controllers consist of the
 administrator. In that list, there are two special controllers:
 MutatingAdmissionWebhook and ValidatingAdmissionWebhook.  These execute the
 mutating and validating (respectively) [admission control
-webhooks](/docs/admin/extensible-admission-controllers/#external-admission-webhooks)
+webhooks](/docs/admin/extensible-admission-controllers.md#external-admission-webhooks)
```
Comment (mistyhacks) [inline content/en/docs/admin/admission-controllers.md]: Update this link and all links that currently start with `/docs/` to the new structure, which starts with `/content/en/docs/`.

File: content/en/docs/admin/admission-controllers.md [hunk 5]
```diff
@@ -60,13 +61,12 @@ admission control plugins:
...
 ```
 
-{{< note >}}
 **Note**: Depending on the way your Kubernetes cluster is deployed and how the
 API server is started, you may need to apply the settings in different ways.
 For example, you may have to modify the systemd unit file if the API server is
 deployed as a systemd service, you may modify the manifest file for the API
 server if Kubernetes is deployed in a self-hosted way.
-{{< /note >}}
+{: .note}
```
Comment (mistyhacks) [inline content/en/docs/admin/admission-controllers.md]: Use Hugo short-codes for this and all admonitions. See the source of https://raw.githubusercontent.com/kubernetes/website/master/content/en/docs/test.md for the syntax, or https://kubernetes.io/docs/test/ for what it looks like. Or just look at the red diff lines because it looks like this admonition was already updated to use the new syntax and your PR undoes it for some reason.

File: content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md [hunk 6]
```diff
@@ -3,23 +3,23 @@ approvers:
...
 
-{{% capture overview %}}
-{{< feature-state for_k8s_version="v1.10" state="beta" >}}
+{% capture overview %}
 
-Persistent volume claims (PVCs) that are in active use by a pod and persistent volumes (PVs) that are bound to PVCs can be protected from pre-mature removal.
+Persistent volume claims (PVCs) that are in active use by a pod and persistent volumes (PVs) that are bound to PVCs can be protected from premature removal.
 
-{{% /capture %}}
+{% endcapture %}
```
Comment (mistyhacks) [inline content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md]: /assign @chenopis 

Which capture syntax is the right one now? I suspect Hugo fixed this and then changes in this PR are effectively undoing the fix.

File: content/en/docs/admin/admission-controllers.md [hunk 7]
```diff
@@ -143,6 +143,8 @@ enabling this admission controller.
 This admission controller mitigates the problem where the API server gets flooded by
 event requests. The cluster admin can specify event rate limits by:
 
+ * Ensuring that `eventratelimit.admission.k8s.io/v1alpha1=true` is included in the
```
Comment (msau42) [inline content/en/docs/admin/admission-controllers.md]: Is this related to the StorageProtection feature?

**Comment by mistyhacks** [comment]
By the way, I was wrong about the need to change `/docs/` to `/content/en/docs/` so you can ignore the request to do that. The things about Hugo shortcodes still need to be addressed.

File: content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md (hunk 6 above)
Comment (chenopis) [reply content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md]: Post-Hugo, it's the `{{ }}` tags, so in this case it's `{{% /capture %}}`. Yeah, it looks like this PR is reverting the changes needed for Hugo.

**Comment by mistyhacks** [comment]
I was wrong about the need to change /docs/ to /content/en/docs/ so you can ignore the request to do that. The things about Hugo shortcodes still need to be addressed.

**Comment by chenopis** [comment]
@msau42 Unfortunately, since there's always a minor Kubernetes release in flight, there was never a perfect time to do the Hugo migration. We did it as early in the release cycle as possible to minimize the number of open 1.11 PRs that would be affected. Also, because the liquid template syntax `{% %}` was embedded in the markdown, there was no good way to separate out the Hugo changes from the content changes. We ran an auto-conversion script on the repo, but unmerged PRs have to be dealt w/ on a one-off basis.

File: content/en/docs/admin/admission-controllers.md [hunk 8]
```diff
@@ -9,7 +9,8 @@ reviewers:
 title: Using Admission Controllers
 ---
 
-{{< toc >}}
```
Comment (mistyhacks) [inline content/en/docs/admin/admission-controllers.md]: This needs to be reverted.

File: content/en/docs/concepts/storage/persistent-volumes.md [hunk 9]
```diff
@@ -9,7 +9,8 @@ title: Persistent Volumes
 
 This document describes the current state of `PersistentVolumes` in Kubernetes. Familiarity with [volumes](/docs/concepts/storage/volumes/) is suggested.
 
-{{< toc >}}
```
Comment (mistyhacks) [inline content/en/docs/concepts/storage/persistent-volumes.md]: This needs to be reverted.

File: content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md [hunk 10]
```diff
@@ -3,23 +3,23 @@ approvers:
 - msau42
 - jsafrane
 title: Storage Object in Use Protection
-content_template: templates/task
```
Comment (mistyhacks) [inline content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md]: This needs to be reverted.

File: content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md (hunk 6 above)
Comment (mistyhacks) [reply content/en/docs/tasks/administer-cluster/storage-object-in-use-protection.md]: Cool, so these lines need to be reverted to use the Hugo syntax.

**Comment by msau42** [comment]
@chenopis can we not submit the hugo changes in another PR, so that this one only contains the content related to the feature?

**Comment by mistyhacks** [comment]
The Hugo stuff is already fixed in the `master` and `release-1.11` branches. The changeset in this PR changes some of those fixes back, because the PR modifies the same lines of code. So yes, the changes can be separated out by reverting those parts back to the state they are already in the `release-1.11` branch. That is my suggested fix for all of the changes above. For instance, the PR changes admonitions whose underlying structure was changed by the Hugo migration, and the PR now needs to take that into account.

The other option is for the author to close this PR and redo the changes manually, but that seems harder to me than just fixing them here. It is of course up to them.

**Comment by mistyhacks** [comment]
I'll try to fix the Hugo stuff to move this along. I think this PR represents two different line items in the 1.11 release spreadsheet so I would like this PR to be in good shape.

File: content/en/docs/admin/admission-controllers.md [hunk 11]
```diff
@@ -563,15 +564,14 @@ See the [resourceQuota design doc](https://git.k8s.io/community/contributors/des
...
-This admission controller implements automation for [serviceAccounts](/docs/tasks/configure-pod-container/configure-service-account/).
+This admission controller implements automation for [serviceAccounts](/docs/user-guide/service-accounts).
 We strongly recommend using this admission controller if you intend to make use of Kubernetes `ServiceAccount` objects.
 
-### Storage Object in Use Protection (beta)
-{{< feature-state for_k8s_version="v1.10" state="beta" >}}
+### Storage Object in Use Protection
 The `StorageObjectInUseProtection` plugin adds the `kubernetes.io/pvc-protection` or `kubernetes.io/pv-protection` finalizers to newly created Persistent Volume Claims (PVCs) or Persistent Volumes (PV). In case a user deletes a PVC or PV the PVC or PV is not removed until the finalizer is removed from the PVC or PV by PVC or PV Protection Controller. Refer to the [Storage Object in Use Protection](/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection) for more detailed information.
 
 ### ValidatingAdmissionWebhook (alpha in 1.8; beta in 1.9)
```
Comment (mistyhacks) [inline content/en/docs/admin/admission-controllers.md]: This comment is really about lines 610++ below, but I can't add a comment there. All of that content relates to versions of Kubernetes that are no longer supported when 1.11 is released (only 1.7+ will be supported), so should they just be removed in the `release-1.11` branch (aka this commit)?

**Comment by Bradamant3** [comment]
This preview link is good for current state of changes, but it looks to me as though changes are still needed. HTH:
https://deploy-preview-8159--kubernetes-io-vnext-staging.netlify.com
(according to Steve, you should be able to reload to include more recent commits. I'll double-check against other deploys.)

**Comment by pospispa** [comment]
@mistyhacks thank you for the Hugo changes. I rebased the PR so that it doesn't contain any Hugo changes.

@Bradamant3 please, let me know whether you request any additional changes or whether I can squash the commits.

**Comment by msau42** [comment]
Thanks, this looks a lot better now and much easier to review!
/lgtm

**Comment by mistyhacks** [comment]
Rebased once more and now this looks good to me.

/lgtm
/approve
